File: Rezoom.Test/TestExceptionFinally.fs
module Rezoom.Test.TestExceptionFinally
open System
open Rezoom
open NUnit.Framework
open FsUnit

[<Test>]
let ``finally no throw`` () =
    let mutable ran = 0
    {   Task = fun () ->
            plan {
                try
                    let! q = send "q"
                    let! r = send "r"
                    return q + r
                finally
                    ran <- ran + 1
            }
        Batches =
            [   [ "q" ]
                [ "r" ]
            ]
        Result = Good "qr"
    } |> test
    if ran <> 1 then
        failwithf "ran was %d" ran

[<Test>]
let ``simple finally`` () =
    let mutable ran = 0
    {   Task = fun () ->
            plan {
                try
                    explode "fail"
                    return 2
                finally
                    ran <- ran + 1
            }
        Batches = []
        Result = Bad (fun _ -> ran = 1)
    } |> test

[<Test>]
let ``bound finally`` () =
    let mutable ran = 0
    {   Task = fun () ->
            plan {
                try
                    let! x = send "x"
                    explode "fail"
                    return 2
                finally
                    ran <- ran + 1
            }
        Batches = [["x"]]
        Result = Bad (fun _ -> ran = 1)
    } |> test

[<Test>]
let ``finally failing prepare`` () =
    let mutable ran = 0
    {   Task = fun () ->
            plan {
                try
                    let! x = failingPrepare "fail" "x"
                    return 2
                finally
                    ran <- ran + 1
            }
        Batches = []
        Result = Bad (fun _ -> ran = 1)
    } |> test

[<Test>]
let ``finally failing retrieve`` () =
    let mutable ran = 0
    {   Task = fun () ->
            plan {
                try
                    let! x = failingRetrieve "fail" "x"
                    return 2
                finally
                    ran <- ran + 1
            }
        Batches = [["x"]]
        Result = Bad (fun _ -> ran = 1)
    } |> test

[<Test>]
let ``using with throw`` () =
    let mutable ran = 0
    {   Task = fun () ->
            plan {
                use d = { new IDisposable with member x.Dispose() = ran <- ran + 1 }
                let! x = failingRetrieve "fail" "x"
                return 2
            }
        Batches = [["x"]]
        Result = Bad (fun _ -> ran = 1)
    } |> test

[<Test>]
let ``using without throw`` () =
    let mutable ran = 0
    {   Task = fun () ->
            plan {
                use d = { new IDisposable with member x.Dispose() = ran <- ran + 1 }
                let! x = send "x"
                return 2
            }
        Batches = [["x"]]
        Result = Good 2
    } |> test

[<Test>]
let ``nested finally`` () =
    let mutable counter = 0
    let mutable first = 0
    let mutable next = 0
    {   Task = fun () ->
            plan {
                try
                    try
                        let! x = failingRetrieve "fail" "x"
                        return 2
                    finally
                        counter <- counter + 1
                        first <- counter
                finally
                    counter <- counter + 1
                    next <- counter
            }
        Batches = [["x"]]
        Result = Bad (fun _ ->
            counter = 2 && first = 1 && next = 2)
    } |> test

[<Test>]
let ``concurrent retrieval abortion good last`` () =
    let mutable ranFinally = false
    let deadly query =
        plan {
            let! x = failingRetrieve "fail" query
            return x
        }
    let good query =
        plan {
            try
                let! result = send query
                let! next = send "jim"
                return result + next
            finally
                ranFinally <- true
        }
    {   Task = fun () ->
            plan {
                let! x, y, z =
                    deadly "x", deadly "y", good "z"
                return x + y + z
            }
        Batches =
            [   [ "x"; "y"; "z" ]
            ]
        Result = Bad (fun ex -> ranFinally)
    } |> test

[<Test>]
let ``concurrent retrieval abortion good first`` () =
    let mutable ranFinally = false
    let deadly query =
        plan {
            let! x = failingRetrieve "fail" query
            return x
        }
    let good query =
        plan {
            try
                let! result = send query
                let! next = send "jim"
                return result + next
            finally
                ranFinally <- true
        }
    {   Task = fun () ->
            plan {
                let! x, y, z =
                    good "x", deadly "y", deadly "z"
                return x + y + z
            }
        Batches =
            [   [ "x"; "y"; "z" ]
            ]
        Result = Bad (fun ex -> ranFinally)
    } |> test

[<Test>]
let ``concurrent retrieval abortion good middle`` () =
    let mutable ranFinally = false
    let deadly query =
        plan {
            let! x = failingRetrieve "fail" query
            return x
        }
    let good query =
        plan {
            try
                let! result = send query
                let! next = send "jim"
                return result + next
            finally
                ranFinally <- true
        }
    {   Task = fun () ->
            plan {
                let! x, y, z =
                    deadly "x", good "y", deadly "z"
                return x + y + z
            }
        Batches =
            [   [ "x"; "y"; "z" ]
            ]
        Result = Bad (fun ex -> ranFinally)
    } |> test

[<Test>]
let ``concurrent logic abortion good last`` () =
    let mutable ranFinally = false
    let deadly query =
        plan {
            failwith "exn"
            let! x = failingRetrieve "fail" query
            return x
        }
    let good query =
        plan {
            try
                let! result = send query
                let! next = send "jim"
                return result + next
            finally
                ranFinally <- true
        }
    {   Task = fun () ->
            plan {
                let! x, y, z =
                    deadly "x", deadly "y", good "z"
                return x + y + z
            }
        Batches =
            [
            ]
        Result = Bad (fun ex -> ranFinally)
    } |> test

[<Test>]
let ``concurrent logic abortion good first`` () =
    let mutable ranFinally = false
    let deadly query =
        plan {
            failwith "exn"
            let! x = failingRetrieve "fail" query
            return x
        }
    let good query =
        plan {
            try
                let! result = send query
                let! next = send "jim"
                return result + next
            finally
                ranFinally <- true
        }
    {   Task = fun () ->
            plan {
                let! x, y, z =
                    good "x", deadly "y", deadly "z"
                return x + y + z
            }
        Batches =
            [
            ]
        Result = Bad (fun ex -> ranFinally)
    } |> test

[<Test>]
let ``concurrent logic abortion good middle`` () =
    let mutable ranFinally = false
    let deadly query =
        plan {
            failwith "exn"
            let! x = failingRetrieve "fail" query
            return x
        }
    let good query =
        plan {
            try
                let! result = send query
                let! next = send "jim"
                return result + next
            finally
                ranFinally <- true
        }
    {   Task = fun () ->
            plan {
                let! x, y, z =
                    deadly "x", good "y", deadly "z"
                return x + y + z
            }
        Batches =
            [
            ]
        Result = Bad (fun ex -> ranFinally)
    } |> test

[<Test>]
let ``concurrent loop retrieval abortion`` () =
    let mutable ranFinally = false
    let deadly query =
        plan {
            let! x = failingRetrieve "fail" query
            return ()
        }
    let good query =
        plan {
            try
                let! result = send query
                let! next = send "jim"
                return ()
            finally
                ranFinally <- true
        }
    {   Task = fun () ->
            plan {
                for q in batch ["x"; "y"; "z"] do
                    if q = "y" then
                        do! good q
                    else
                        do! deadly q
                return ()
            }
        Batches =
            [   [ "x"; "y"; "z" ]
            ]
        Result = Bad (fun _ ->
            ranFinally)
    } |> test

[<Test>]
let ``concurrent loop logic abortion`` () =
    let mutable ranFinally = false
    let deadly query =
        plan {
            failwith "logic"
            let! x = failingRetrieve "fail" query
            return ()
        }
    let good query =
        plan {
            try
                let! result = send query
                let! next = send "jim"
                return ()
            finally
                ranFinally <- true
        }
    {   Task = fun () ->
            plan {
                for q in batch ["x"; "y"; "z"] do
                    if q = "y" then
                        do! good q
                    else
                        do! deadly q
                return ()
            }
        Batches =
            [
            ]
        Result = Bad (fun _ ->
            ranFinally)
    } |> test
        

File: src/Rezoom.CS.Test/TestExceptionFinally.cs
using System;
using NUnit.Framework;
using Rezoom;
using Rezoom.CS;
using static Rezoom.CS.Test.Helpers;

namespace Rezoom.CS.Test;

[TestFixture]
public class TestExceptionFinally
{
    private sealed class DisposeRecorder : IDisposable
    {
        private readonly Action _onDispose;
        public DisposeRecorder(Action onDispose) { _onDispose = onDispose; }
        public void Dispose() => _onDispose();
    }

    [Test]
    public void Finally_no_throw()
    {
        int ran = 0;
        async Plan<string> Task()
        {
            try
            {
                var q = await Send("q");
                var r = await Send("r");
                return q + r;
            }
            finally { ran++; }
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "q" }, new[] { "r" } },
            ExpectedResult<string>.Ok("qr"));
        Assert.That(ran, Is.EqualTo(1));
    }

    [Test]
    public void Simple_finally()
    {
        int ran = 0;
        async Plan<int> Task()
        {
            try
            {
                Explode("fail");
                return 2;
            }
            finally { ran++; }
        }
        TestRunner.Run(
            Task,
            System.Array.Empty<string[]>(),
            ExpectedResult<int>.Fail(_ => ran == 1));
    }

    [Test]
    public void Bound_finally()
    {
        int ran = 0;
        async Plan<int> Task()
        {
            try
            {
                var x = await Send("x");
                Explode("fail");
                return 2;
            }
            finally { ran++; }
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x" } },
            ExpectedResult<int>.Fail(_ => ran == 1));
    }

    [Test]
    public void Finally_failing_prepare()
    {
        int ran = 0;
        async Plan<int> Task()
        {
            try
            {
                var x = await FailingPrepare("fail", "x");
                return 2;
            }
            finally { ran++; }
        }
        TestRunner.Run(
            Task,
            System.Array.Empty<string[]>(),
            ExpectedResult<int>.Fail(_ => ran == 1));
    }

    [Test]
    public void Finally_failing_retrieve()
    {
        int ran = 0;
        async Plan<int> Task()
        {
            try
            {
                var x = await FailingRetrieve("fail", "x");
                return 2;
            }
            finally { ran++; }
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x" } },
            ExpectedResult<int>.Fail(_ => ran == 1));
    }

    [Test]
    public void Using_with_throw()
    {
        int ran = 0;
        async Plan<int> Task()
        {
            using var d = new DisposeRecorder(() => ran++);
            var x = await FailingRetrieve("fail", "x");
            return 2;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x" } },
            ExpectedResult<int>.Fail(_ => ran == 1));
    }

    [Test]
    public void Using_without_throw()
    {
        int ran = 0;
        async Plan<int> Task()
        {
            using var d = new DisposeRecorder(() => ran++);
            var x = await Send("x");
            return 2;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x" } },
            ExpectedResult<int>.Ok(2));
        Assert.That(ran, Is.EqualTo(1));
    }

    [Test]
    public void Nested_finally()
    {
        int counter = 0, first = 0, next = 0;
        async Plan<int> Task()
        {
            try
            {
                try
                {
                    var x = await FailingRetrieve("fail", "x");
                    return 2;
                }
                finally { counter++; first = counter; }
            }
            finally { counter++; next = counter; }
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x" } },
            ExpectedResult<int>.Fail(_ => counter == 2 && first == 1 && next == 2));
    }

    [Test]
    public void Concurrent_retrieval_abortion_good_last()
    {
        bool ranFinally = false;
        async Plan<string> Deadly(string query) => await FailingRetrieve("fail", query);
        async Plan<string> Good(string query)
        {
            try
            {
                var result = await Send(query);
                var next = await Send("jim");
                return result + next;
            }
            finally { ranFinally = true; }
        }
        async Plan<string> Task()
        {
            var (x, y, z) = await Plans.Tuple(Deadly("x"), Deadly("y"), Good("z"));
            return x + y + z;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x", "y", "z" } },
            ExpectedResult<string>.Fail(_ => ranFinally));
    }

    [Test]
    public void Concurrent_retrieval_abortion_good_first()
    {
        bool started = false;
        bool ranFinally = false;
        async Plan<string> Deadly(string query) => await FailingRetrieve("fail", query);
        async Plan<string> Good(string query)
        {
            started = true;
            try
            {
                var result = await Send(query);
                var next = await Send("jim");
                return result + next;
            }
            finally { ranFinally = true; }
        }
        async Plan<string> Task()
        {
            var (x, y, z) = await Plans.Tuple(Good("x"), Deadly("y"), Deadly("z"));
            return x + y + z;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x", "y", "z" } },
            ExpectedResult<string>.Fail(_ => started && ranFinally));
    }

    [Test]
    public void Concurrent_retrieval_abortion_good_middle()
    {
        bool started = false;
        bool ranFinally = false;
        async Plan<string> Deadly(string query) => await FailingRetrieve("fail", query);
        async Plan<string> Good(string query)
        {
            started = true;
            try
            {
                var result = await Send(query);
                var next = await Send("jim");
                return result + next;
            }
            finally { ranFinally = true; }
        }
        async Plan<string> Task()
        {
            var (x, y, z) = await Plans.Tuple(Deadly("x"), Good("y"), Deadly("z"));
            return x + y + z;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x", "y", "z" } },
            ExpectedResult<string>.Fail(_ => started && ranFinally));
    }

    [Test]
    public void Concurrent_logic_abortion_good_last()
    {
        bool started = false;
        bool ranFinally = false;
        async Plan<string> Deadly(string query)
        {
            throw new Exception("exn");
#pragma warning disable CS0162 // Unreachable code
            var x = await FailingRetrieve("fail", query);
            return x;
#pragma warning restore CS0162
        }
        async Plan<string> Good(string query)
        {
            started = true;
            try
            {
                var result = await Send(query);
                var next = await Send("jim");
                return result + next;
            }
            finally { ranFinally = true; }
        }
        async Plan<string> Task()
        {
            var (x, y, z) = await Plans.Tuple(Deadly("x"), Deadly("y"), Good("z"));
            return x + y + z;
        }
        TestRunner.Run(
            Task,
            System.Array.Empty<string[]>(),
            ExpectedResult<string>.Fail(_ => !started && !ranFinally));
    }

    [Test]
    public void Concurrent_logic_abortion_good_first()
    {
        bool ranFinally = false;
        async Plan<string> Deadly(string query)
        {
            throw new Exception("exn");
#pragma warning disable CS0162
            var x = await FailingRetrieve("fail", query);
            return x;
#pragma warning restore CS0162
        }
        async Plan<string> Good(string query)
        {
            try
            {
                var result = await Send(query);
                var next = await Send("jim");
                return result + next;
            }
            finally { ranFinally = true; }
        }
        async Plan<string> Task()
        {
            var (x, y, z) = await Plans.Tuple(Good("x"), Deadly("y"), Deadly("z"));
            return x + y + z;
        }
        TestRunner.Run(
            Task,
            System.Array.Empty<string[]>(),
            ExpectedResult<string>.Fail(_ => ranFinally));
    }

    [Test]
    public void Concurrent_logic_abortion_good_middle()
    {
        bool ranFinally = false;
        async Plan<string> Deadly(string query)
        {
            throw new Exception("exn");
#pragma warning disable CS0162
            var x = await FailingRetrieve("fail", query);
            return x;
#pragma warning restore CS0162
        }
        async Plan<string> Good(string query)
        {
            try
            {
                var result = await Send(query);
                var next = await Send("jim");
                return result + next;
            }
            finally { ranFinally = true; }
        }
        async Plan<string> Task()
        {
            var (x, y, z) = await Plans.Tuple(Deadly("x"), Good("y"), Deadly("z"));
            return x + y + z;
        }
        TestRunner.Run(
            Task,
            System.Array.Empty<string[]>(),
            ExpectedResult<string>.Fail(_ => ranFinally));
    }

    [Test]
    public void Concurrent_loop_retrieval_abortion()
    {
        bool started = false;
        bool ranFinally = false;
        async Plan<int> Deadly(string query) { _ = await FailingRetrieve("fail", query); return 0; }
        async Plan<int> Good(string query)
        {
            started = true;
            try
            {
                _ = await Send(query);
                _ = await Send("jim");
                return 0;
            }
            finally { ranFinally = true; }
        }
        async Plan<int> Task()
        {
            await Plans.ForBatch(new[] { "x", "y", "z" }, async q =>
            {
                if (q == "y") return await Good(q);
                return await Deadly(q);
            });
            return 0;
        }
        TestRunner.Run(
            Task,
            new[] { new[] { "x", "y", "z" } },
            ExpectedResult<int>.Fail(_ => started && ranFinally));
    }

    [Test]
    public void Concurrent_loop_logic_abortion()
    {
        bool ranFinally = false;
        async Plan<int> Deadly(string query)
        {
            throw new Exception("logic");
#pragma warning disable CS0162
            _ = await FailingRetrieve("fail", query);
            return 0;
#pragma warning restore CS0162
        }
        async Plan<int> Good(string query)
        {
            try
            {
                _ = await Send(query);
                _ = await Send("jim");
                return 0;
            }
            finally { ranFinally = true; }
        }
        async Plan<int> Task()
        {
            await Plans.ForBatch(new[] { "x", "y", "z" }, async q =>
            {
                if (q == "y") return await Good(q);
                return await Deadly(q);
            });
            return 0;
        }
        TestRunner.Run(
            Task,
            System.Array.Empty<string[]>(),
            ExpectedResult<int>.Fail(_ => ranFinally));
    }
}
